### PR TITLE
Lucene DocId to PinotDocId cache

### PR DESCRIPTION
**Pull request by siddharthteotia**
Cache the lucene docID to pinot docID mapping per text index. This has to be thread safe since it is shared by all threads (query threads) reading the index.

Storing this mapping on-heap structures (concurrent hash map, atomic int array etc) will add significant heap overhead.  Take an example of 10million docs per segment and we store a 4 byte doc id. This leads to 40MB per indexes. If there are 100 text indexes across segments on a server, we are looking at 4GB of heap overhead.

So the mapping has to be in off-heap memory (direct or mmap'd). The current solution builds the mapping upfront during segment load in a memory mapped file. During query processing, we avoid all calls to indexSearcher.doc(int docID). 

We had noticed in our internal perf testing that as QPS is increased, performance degrades. The docID converter (lucene to Pinot) is a CPU hogger (confirmed through FlameGraphs in YourKit) and was becoming a bottleneck for QPS based testing. On the other hand, when running the same set of expensive queries as one-off, the execution time of the each query was drastically improved. 

This change should provide significant improvements for throughput testing.  Verified by running PerfBenchmarkRunner without and without the changes.

### CODE REVIEW

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java [hunk 1]
```diff
@@ -148,8 +153,12 @@ private MutableRoaringBitmap getPinotDocIds(MutableRoaringBitmap luceneDocIds) {
     try {
       while (luceneDocIDIterator.hasNext()) {
         int luceneDocId = luceneDocIDIterator.next();
-        Document document = _indexSearcher.doc(luceneDocId);
-        int pinotDocId = Integer.valueOf(document.get(LuceneTextIndexCreator.LUCENE_INDEX_DOC_ID_COLUMN_NAME));
+        Integer pinotDocId = _luceneDocIDToPinotDocIDCache.get(luceneDocId);
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java]: Please use `computeIfAbsent` instead

Comment (siddharthteotia) [reply pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java]: I am not using this method anymore. The mapping is built once during segment load and memory mapped.

**Comment by siddharthteotia** [comment]
@Jackie-Jiang , @kishoreg please take a look at this. I'd like to provide a new build for ongoing internal perf testing asap.

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java (hunk 1 above)
Comment (siddharthteotia) [reply pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java]: Also added a TODO to consider building mapping on-the-fly during query processing -- we can have two methods and make it configurable based on number of docs per segment.

File: pinot-core/src/test/java/org/apache/pinot/queries/TestTextSearchQueries.java [hunk 2]
```diff
@@ -89,7 +91,7 @@
   private RecordReader _recordReader;
   Schema _schema;
 
-  private List<IndexSegment> _indexSegments = new ArrayList<>(1);
+  private static List<IndexSegment> _indexSegments = new ArrayList<>(1);
```
Comment (Jackie-Jiang) [inline pinot-core/src/test/java/org/apache/pinot/queries/TestTextSearchQueries.java]: Revert the change in this file?
Also suggesting rename the test to TextSearchQueriesTest for naming convention.

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java [hunk 3]
```diff
@@ -51,6 +54,9 @@
   private final IndexSearcher _indexSearcher;
   private final QueryParser _queryParser;
   private final String _column;
+  private DocIdReaderWriter _docIdReaderWriter;
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java]: `private final`

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java [hunk 4]
```diff
@@ -70,6 +76,10 @@ public LuceneTextIndexReader(String column, File segmentIndexDir) {
       // Disable Lucene query result cache. While it helps a lot with performance for
       // repeated queries, on the downside it cause heap issues.
       _indexSearcher.setQueryCache(null);
+      // TODO: consider using a threshold of num docs per segment to decide between building
+      // mapping file upfront on segment load v/s on-the-fly during query processing
+      _docIdReaderWriter = new DocIdReaderWriter(segmentIndexDir, _column, numDocs);
+      _docIdReaderWriter.buildDocIdMapping(numDocs);
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java]: We build the mapping every time we load the index? You saved the mapping into a file right?

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java [hunk 5]
```diff
@@ -169,5 +178,49 @@ public void close()
...
+    private PinotDataBuffer _buffer;
+
+    DocIdReaderWriter(File segmentIndexDir, String column, int numDocs) throws Exception {
+      int length = Integer.BYTES * numDocs;
+      File docIdMappingFile = new File(SegmentDirectoryPaths.findSegmentDirectory(segmentIndexDir),
+          column + LUCENE_TEXT_INDEX_DOCID_MAPPING_FILE_EXTENSION);
+      // For newly added segments, this file will not exist.
+      // For segment refresh, segment reload and server restart, file will exist,
+      // but we don't know if we are here for refresh v/s reload v/s restart.
+      // In case of refresh, we have to build the mapping again, but in case of
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java]: For segment refresh, this file should not exist as we delete the old segment

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java [hunk 6]
```diff
@@ -169,5 +178,49 @@ public void close()
...
+      // reload and restart, we don't. Also, reload has a sub-case where this text index
+      // was indeed created during reload (user enabled on existing or newly added column).
+      // Since there is no way to distinguish why we are here, we build the mapping again
+      // regardless.
+      // TODO: see if we can prefetch the pages
+      _buffer =
+          PinotDataBuffer.mapFile(docIdMappingFile, false, 0, length, ByteOrder.BIG_ENDIAN, getClass().getSimpleName());
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java]: For better performance, I would probably choose native order as this index is always local to one instance?

Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java]: Please include the column name into the description (last argument) to distinguish different columns

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java [hunk 7]
```diff
@@ -169,5 +178,49 @@ public void close()
...
+    }
+
+    public void buildDocIdMapping(int numDocs) {
+      for (int i = 0; i < numDocs; i++) {
+        try {
+          Document document = _indexSearcher.doc(i);
+          int pinotDocId = Integer.valueOf(document.get(LuceneTextIndexCreator.LUCENE_INDEX_DOC_ID_COLUMN_NAME));
+          _buffer.putInt(i * Integer.BYTES, pinotDocId);
+        } catch (Exception e) {
+          LOGGER.error("Failed to build doc id mapping during segment load for column:{},docID:{},error:{}. Will continue and build mapping on the fly",
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java]: Throw this exception out instead of logging an ERROR. If this step fails, JVM will crash when reading the buffer.

File: pinot-core/src/test/java/org/apache/pinot/queries/TestTextSearchQueries.java (hunk 2 above)
Comment (siddharthteotia) [reply pinot-core/src/test/java/org/apache/pinot/queries/TestTextSearchQueries.java]: oops. Sorry, this was due to local debugging. Will remove and also reformat

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java (hunk 4 above)
Comment (siddharthteotia) [reply pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java]: We check for existence and return.

Comment (siddharthteotia) [reply pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java]: Also, I explored doing this in TextIndexHandler. But that is not good since it requires to open the lucene index, create searcher twice (both in handler and here anyway for query processing). I think it is better to avoid that and just open the lucene index reader and search just once per index.

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java (hunk 3 above)
Comment (siddharthteotia) [reply pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java]: done

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java (hunk 5 above)
Comment (siddharthteotia) [reply pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java]: Yes, that's what I stated in the comment.

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java (hunk 6 above)
Comment (siddharthteotia) [reply pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java]: Yes, as discussed using Little Endian and indicated in the comments too. We are unlikely to run on Solaris/Sparc so using LE is fine which is the case on Linux/x86

Comment (siddharthteotia) [reply pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java]: done

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java (hunk 7 above)
Comment (siddharthteotia) [reply pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java]: done

**Comment by siddharthteotia** [comment]
@Jackie-Jiang , I have addressed the review comments. Please take another look

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java [hunk 8]
```diff
@@ -169,5 +178,57 @@ public void close()
       throws IOException {
     _indexReader.close();
     _indexDirectory.close();
+    _docIdReaderWriter.close();
+  }
+
+  private class DocIdReaderWriter implements Closeable {
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java]: `private static class`

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java [hunk 9]
```diff
@@ -169,5 +178,57 @@ public void close()
...
+            _column + getClass().getSimpleName());
+      }
+    }
+
+    public void buildDocIdMapping(int numDocs) {
+      if (!_mappingExists) {
+        for (int i = 0; i < numDocs; i++) {
+          try {
+            Document document = _indexSearcher.doc(i);
+            int pinotDocId = Integer.valueOf(document.get(LuceneTextIndexCreator.LUCENE_INDEX_DOC_ID_COLUMN_NAME));
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java]: `Integer.parseInt()`

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java [hunk 10]
```diff
@@ -169,5 +178,57 @@ public void close()
...
+        _buffer = PinotDataBuffer.mapFile(docIdMappingFile, /* readOnly */ true, 0, length, ByteOrder.LITTLE_ENDIAN,
+            _column + getClass().getSimpleName());
+      } else {
+        _mappingExists = false;
+        _buffer = PinotDataBuffer.mapFile(docIdMappingFile, /* readOnly */ false, 0, length, ByteOrder.LITTLE_ENDIAN,
+            _column + getClass().getSimpleName());
+      }
+    }
+
+    public void buildDocIdMapping(int numDocs) {
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java]: Merge this into the constructor, no need to track an extra boolean.

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java [hunk 11]
```diff
@@ -60,7 +66,7 @@
    * @param column column name
    * @param segmentIndexDir segment index directory
    */
-  public LuceneTextIndexReader(String column, File segmentIndexDir) {
+  public LuceneTextIndexReader(String column, File segmentIndexDir, int numDocs) {
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java]: Recommend making second argument `indexDir` to denote that it is top level dir

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java [hunk 12]
```diff
@@ -121,7 +131,7 @@ public MutableRoaringBitmap getDocIds(Object value) {
     try {
       Query query = _queryParser.parse(searchQuery);
       _indexSearcher.search(query, docIDCollector);
-      return getPinotDocIds(docIDs);
+      return getPinotDocIdsFromMappingFile(docIDs);
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java]: I prefer the original name

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java [hunk 13]
```diff
@@ -121,7 +131,7 @@ public MutableRoaringBitmap getDocIds(Object value) {
...
     } catch (Exception e) {
       LOGGER.error("Failed while searching the text index for column {}, search query {}, exception {}", _column,
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java]: You don't really need to log the error if you are going to throw out the exception. The catcher will log it with the stack trace

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java [hunk 14]
```diff
@@ -142,18 +152,17 @@ public MutableRoaringBitmap getDocIds(Object value) {
...
       while (luceneDocIDIterator.hasNext()) {
         int luceneDocId = luceneDocIDIterator.next();
-        Document document = _indexSearcher.doc(luceneDocId);
-        int pinotDocId = Integer.valueOf(document.get(LuceneTextIndexCreator.LUCENE_INDEX_DOC_ID_COLUMN_NAME));
+        int pinotDocId = _docIdReaderWriter.getInt(luceneDocId * Integer.BYTES);
         actualDocIDs.add(pinotDocId);
       }
     } catch (Exception e) {
-      throw new RuntimeException("Error: failed while retrieving document from index: " + e);
+      throw new RuntimeException("Error: failed while retrieving pinot doc id from mapping file: " + e);
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java]: No need to catch, you can directly throw the exception

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java [hunk 15]
```diff
@@ -169,5 +178,57 @@ public void close()
...
+      // The mapping is local to a segment. It is created on the server during segment load.
+      // Unless we are running Pinot on Solaris/SPARC, the underlying architecture is
+      // LITTLE_ENDIAN (Linux/x86). So use that as byte order.
+      if (docIdMappingFile.exists()) {
+        // we will be here for segment reload and server restart
+        // for refresh, we will not be here since segment is deleted/replaced
+        // TODO: see if we can prefetch the pages
+        _mappingExists = true;
+        _buffer = PinotDataBuffer.mapFile(docIdMappingFile, /* readOnly */ true, 0, length, ByteOrder.LITTLE_ENDIAN,
+            _column + getClass().getSimpleName());
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java]: `"Text index docId mapping buffer: " + _column`

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java [hunk 16]
```diff
@@ -169,5 +178,57 @@ public void close()
...
+            int pinotDocId = Integer.valueOf(document.get(LuceneTextIndexCreator.LUCENE_INDEX_DOC_ID_COLUMN_NAME));
+            _buffer.putInt(i * Integer.BYTES, pinotDocId);
+          } catch (Exception e) {
+            throw new RuntimeException("Failed to build doc id mapping during segment load: " + e);
+          }
+        }
+      }
+    }
+
+    int getInt(int offset) {
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java]: `int getPinotDocId(int luceneDocId)` for better abstraction

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java (hunk 8 above)
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java]: I would recommend renaming it to `DocIdTranslator`

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java [hunk 17]
```diff
@@ -169,5 +178,57 @@ public void close()
...
+          column + LUCENE_TEXT_INDEX_DOCID_MAPPING_FILE_EXTENSION);
+      // The mapping is local to a segment. It is created on the server during segment load.
+      // Unless we are running Pinot on Solaris/SPARC, the underlying architecture is
+      // LITTLE_ENDIAN (Linux/x86). So use that as byte order.
+      if (docIdMappingFile.exists()) {
+        // we will be here for segment reload and server restart
+        // for refresh, we will not be here since segment is deleted/replaced
+        // TODO: see if we can prefetch the pages
+        _mappingExists = true;
+        _buffer = PinotDataBuffer.mapFile(docIdMappingFile, /* readOnly */ true, 0, length, ByteOrder.LITTLE_ENDIAN,
```
Comment (kishoreg) [inline pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java]: Can we reuse FixedByteReaderWriter?

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java [hunk 18]
```diff
@@ -169,5 +178,57 @@ public void close()
...
+
+    public void buildDocIdMapping(int numDocs) {
+      if (!_mappingExists) {
+        for (int i = 0; i < numDocs; i++) {
+          try {
+            Document document = _indexSearcher.doc(i);
+            int pinotDocId = Integer.valueOf(document.get(LuceneTextIndexCreator.LUCENE_INDEX_DOC_ID_COLUMN_NAME));
+            _buffer.putInt(i * Integer.BYTES, pinotDocId);
+          } catch (Exception e) {
+            throw new RuntimeException("Failed to build doc id mapping during segment load: " + e);
```
Comment (kishoreg) [inline pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java]: , e

File: pinot-core/src/test/java/org/apache/pinot/queries/TextSearchQueriesTest.java [hunk 19]
```diff
@@ -18,6 +18,7 @@
  */
 package org.apache.pinot.queries;
 
+import com.google.common.base.Stopwatch;
```
[the remark this answers is not in the document]
Comment (siddharthteotia) [reply pinot-core/src/test/java/org/apache/pinot/queries/TextSearchQueriesTest.java]: Done

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/loader/invertedindex/TextIndexHandler.java [hunk 20]
```diff
@@ -154,6 +152,11 @@ private void createTextIndexForColumn(ColumnMetadata columnMetadata)
     int numDocs = columnMetadata.getTotalDocs();
     LOGGER.info("Creating new text index for column: {} in segment: {}", column, _segmentName);
     File segmentIndexDir = SegmentDirectoryPaths.segmentDirectoryFor(_indexDir, _segmentVersion);
+    // The handlers are always invoked by the preprocessor. Before this ImmutableSegmentLoader would have already
```
[the remark this answers is not in the document]
Comment (siddharthteotia) [reply pinot-core/src/main/java/org/apache/pinot/core/segment/index/loader/invertedindex/TextIndexHandler.java]: Done. Wasn't aware of this.

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/loader/invertedindex/TextIndexHandler.java [hunk 21]
```diff
@@ -61,8 +61,26 @@
...
+ * detects the existence of text index and returns.
+ *
+ * (3) A reload is issued on an existing segment after text index is enabled on an existing
+ * column. Read the forward index to create text index.
+ *
+ * (4) A reload is issued on an existing segment after text index is enabled on a newly
+ * added column. In this case, the default column handler would have taken care of adding
+ * forward index for the new column. Read the forward index to create text index.
+ */
 public class TextIndexHandler {
```
[the remark this answers is not in the document]
Comment (siddharthteotia) [reply pinot-core/src/main/java/org/apache/pinot/core/segment/index/loader/invertedindex/TextIndexHandler.java]: done

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java [hunk 22]
```diff
@@ -169,5 +178,57 @@ public void close()
       throws IOException {
     _indexReader.close();
     _indexDirectory.close();
+    _docIdReaderWriter.close();
+  }
+
+  private class DocIdReaderWriter implements Closeable {
+    private PinotDataBuffer _buffer;
```
[the remark this answers is not in the document]
Comment (siddharthteotia) [reply pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java]: done

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java (hunk 8 above)
Comment (siddharthteotia) [reply pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java]: done

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java (hunk 9 above)
Comment (siddharthteotia) [reply pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java]: done

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java (hunk 10 above)
Comment (siddharthteotia) [reply pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java]: done

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java (hunk 11 above)
Comment (siddharthteotia) [reply pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java]: done

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java (hunk 12 above)
Comment (siddharthteotia) [reply pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java]: ohh yes, I was initially thinking of implementing two solutions. Reverted

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java (hunk 13 above)
Comment (siddharthteotia) [reply pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java]: Removed the e.getMessage() from log. The log is still needed though since I want to capture column name and search expression for the failed query

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java (hunk 14 above)
Comment (siddharthteotia) [reply pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java]: done

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java (hunk 15 above)
Comment (siddharthteotia) [reply pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java]: done

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java (hunk 16 above)
Comment (siddharthteotia) [reply pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java]: done

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java (hunk 8 above)
Comment (siddharthteotia) [reply pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java]: done

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/converter/SegmentV1V2ToV3FormatConverter.java [hunk 23]
```diff
@@ -241,6 +242,22 @@ public boolean accept(File dir, String name) {
         Files.copy(indexFile.toPath(), v3LuceneIndexFile.toPath());
       }
     }
+    // if segment reload is issued asking for up-conversion of
```
[the remark this answers is not in the document]
Comment (siddharthteotia) [reply pinot-core/src/main/java/org/apache/pinot/core/segment/index/converter/SegmentV1V2ToV3FormatConverter.java]: Lucene index files (multiple files in a single directory created by Lucene) can't be copied as done in copyForwardIndex

`PinotDataBuffer oldBuffer = reader.getIndexFor(column, indexType);
    long oldBufferSize = oldBuffer.size();
    PinotDataBuffer newBuffer = writer.newIndexFor(column, indexType, oldBufferSize);
    oldBuffer.copyTo(0, newBuffer, 0, oldBufferSize);`

The only thing that can be copied as buffer is the mapping file that we create. I just wanted to keep everything related to lucene in a single method. I can explore how to clean this up.

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java (hunk 17 above)
Comment (siddharthteotia) [reply pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java]: I explored existing reader/writer. Didn't use for the following reasons:

FixedByteChunkReaderWriter uses direct memory (ByteBuffer.allocateDirect). Memory mapping is better.

The FixedBitIntReaderWriter (used for dictionary encoded forward index) has additional bells and whistles due to bit-packing. The return type is an array. I wanted to keep this simple since we use it in exactly one place in lucene code.

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java (hunk 18 above)
Comment (siddharthteotia) [reply pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java]: done

**Comment by siddharthteotia** [comment]
@Jackie-Jiang , @kishoreg 
Addressed the review comments.

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/converter/SegmentV1V2ToV3FormatConverter.java (hunk 23 above)
[the remark this answers is not in the document]
Comment (siddharthteotia) [reply pinot-core/src/main/java/org/apache/pinot/core/segment/index/converter/SegmentV1V2ToV3FormatConverter.java]: Added a TODO -- I have couple of more PRs coming up for text. Will address this

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java [hunk 24]
```diff
@@ -123,8 +133,7 @@ public MutableRoaringBitmap getDocIds(Object value) {
       _indexSearcher.search(query, docIDCollector);
       return getPinotDocIds(docIDs);
     } catch (Exception e) {
-      LOGGER.error("Failed while searching the text index for column {}, search query {}, exception {}", _column,
-          searchQuery, e.getMessage());
+      LOGGER.error("Failed while searching the text index for column {}, search query {},", _column, searchQuery);
       throw new RuntimeException(e);
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java]:
```suggestion
      throw new RuntimeException("Caught exception while searching the text index column: " + _column + " with query: " + searchQuery, e);
```

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java [hunk 25]
```diff
@@ -169,5 +173,50 @@ public void close()
...
+            int pinotDocId = Integer.parseInt(document.get(LuceneTextIndexCreator.LUCENE_INDEX_DOC_ID_COLUMN_NAME));
+            _buffer.putInt(i * Integer.BYTES, pinotDocId);
+          } catch (Exception e) {
+            throw new RuntimeException("Failed to build doc id mapping during segment load, " + e);
+          }
+        }
+      }
+    }
+
+    int getPinotDocId(int offset) {
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java]: Pass in `luceneDocId` and wrap the `_buffer.getInt(luceneDocId * Integer.BYTES)` logic inside

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java [hunk 26]
```diff
@@ -169,5 +173,50 @@ public void close()
...
+      } else {
+        _buffer =
+            PinotDataBuffer.mapFile(docIdMappingFile, /* readOnly */ false, 0, length, ByteOrder.LITTLE_ENDIAN, desc);
+        for (int i = 0; i < numDocs; i++) {
+          try {
+            Document document = indexSearcher.doc(i);
+            int pinotDocId = Integer.parseInt(document.get(LuceneTextIndexCreator.LUCENE_INDEX_DOC_ID_COLUMN_NAME));
+            _buffer.putInt(i * Integer.BYTES, pinotDocId);
+          } catch (Exception e) {
+            throw new RuntimeException("Failed to build doc id mapping during segment load, " + e);
```
[the remark this answers is not in the document]
Comment (siddharthteotia) [reply pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java]: done

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java [hunk 27]
```diff
@@ -123,8 +133,7 @@ public MutableRoaringBitmap getDocIds(Object value) {
       _indexSearcher.search(query, docIDCollector);
       return getPinotDocIds(docIDs);
     } catch (Exception e) {
-      LOGGER.error("Failed while searching the text index for column {}, search query {}, exception {}", _column,
-          searchQuery, e.getMessage());
+      LOGGER.error("Failed while searching the text index for column {}, search query {},", _column, searchQuery);
```
[the remark this answers is not in the document]
Comment (siddharthteotia) [reply pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java]: used in message of RuntimeException

File: pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java (hunk 25 above)
Comment (siddharthteotia) [reply pinot-core/src/main/java/org/apache/pinot/core/segment/index/readers/text/LuceneTextIndexReader.java]: Good catch. Done
